### ISSUE TITLE
Optionally align subqueries to end-time of query_range

### ISSUE THREAD

**Issue by PrestonTaylor** [prometheus/prometheus#9767]
I frequently find myself using offsets to compare previous days/weeks/months to current to get a better idea of how seasonal data is behaving. 

I have run into a roadblock though in going deeper in that I can't figure out how to work with multiple offsets using aggregating functions to get something more 'intelligent' to use as a comparative baseline or alert threshold etc. For example if I wanted to find out what the average Wednesday looks like. 

Some other parts of PromQL inspired this syntax which is certainly not the important piece of this feature request and no doubt may clash with something but seeing it might help explain what I'm thinking.

```
avg(
    (http_requests_total offset [1d:1d:28d])
)
```
This would be similar to in splunk querying 28 days of data with a `timewrap span=1d` applied. So excuse my poor choice of words here, I'm sure there are better but the values in the brackets mean this: [ window length : window separation : how far back to look ] 

This would be seriously helpful in working with seasonal data.

Because this would obviously return multiple series, I assume each one could have a label added to indicate which window it belongs to.

Perhaps there is already some other way to accomplish this in Prometheus, if there is please tell me, I'd love to take advantage.

**Comment by roidelapluie** [comment]
Yes, this would be useful. I am using very complex queries to get this. As you said, I am unsure about the syntax.

**Comment by roidelapluie** [comment]
Your usecase should be possible today with `avg_over_time(rate(metric[5m])[28d:1d])`, but I am not sure how efficient it would be.

**Comment by roidelapluie** [comment]
Your usecase will be possible with `avg_over_time(rate(metric[5m])[28d::1d])`, but requires #9114

**Comment by beorn7** [comment]
Indeed, I think #9114 will enable quite a few things. @LeviHarrison feel encouraged to pick it up again! (o:

**Comment by roidelapluie** [comment]
I have marked #9114 as fixing this issue.

**Comment by PrestonTaylor** [comment]
@roidelapluie just revisiting this out of curiosity, I noticed you mentioned complex queries to solve the same need, I've just been using something like this

```
#foo average day
(
	(
		foo{} offset 1d 
	) + on (bar)
	(
		foo{} offset 2d
	) + on (bar)
	(
		foo{} offset 3d
	)
)/3
```

You wouldn't happen to have anything more flexible even if it's complex? I'm looking to expand this to get 30+ days of data and with a 20 line subquery it's getting ugly not to mention causing problems with the grafana plugin.

**Comment by bboreham** [comment]
I had no idea what "timewrap" meant.  From context I guess it is a Splunk feature.
Based on the PR I think "Optionally align subqueries to end-time of query_range" is better, but maybe this issue is more general?

**Comment by beorn7** [comment]
Hello from the bug scrub!

This is still relevant, and #9114 still needs a new champion. We've changed the title, but we also considered that #9114 could in fact not solve all the use cases raised here.

**Comment by ndk** [comment]
> Hello from the bug scrub!
> 
> This is still relevant, and [#9114](https://github.com/prometheus/prometheus/pull/9114) still needs a new champion. We've changed the title, but we also considered that [#9114](https://github.com/prometheus/prometheus/pull/9114) could in fact not solve all the use cases raised here.

I have some ideas. :)
https://github.com/prometheus/prometheus/pull/17106
